### PR TITLE
raft: refactor inflight

### PR DESCRIPTION
**Pull request by xiang90**
Fix https://github.com/coreos/etcd/issues/4899

/cc @bdarnell @ngaut @ongardie @superfell

### CODE REVIEW

File: raft/progress.go [hunk 1]
```diff
@@ -64,12 +64,17 @@ type Progress struct {
 	RecentActive bool
 
 	// inflights is a sliding window for the inflight messages.
+	// Each inflight message might contain more than one entries
+	// and has a size limit defined in raft config as MaxSizePerMsg.
```
Comment (ongardie) [inline raft/progress.go]: A size limit in number of bytes or number of entries? I'm again confused about the purpose of this whole thing. Are the number of bytes tracked anywhere?

File: raft/progress.go [hunk 2]
```diff
@@ -235,7 +240,11 @@ func (in *inflights) freeFirstOne() { in.freeTo(in.buffer[in.start]) }
 
 // full returns true if the inflights is full.
 func (in *inflights) full() bool {
-	return in.count == in.size
+	return in.count == len(in.buffer)
+}
+
+func (in *inflights) size() int {
```
Comment (ongardie) [inline raft/progress.go]: Might as well inline this.

File: raft/progress.go (hunk 1 above)
Comment (xiang90) [inline raft/progress.go]: Sorry... So a message contains one or multiple raft entries. Each message has a max size.

Say leader sends 8 entries with 1MB per entries. For each message, the max size is 4MB. With flow control, two messages with entries [1,2,3,4] and [5,6,7,8] will be sent. We will have [4, 8] in inflights.

Once follower replies with index 8, then we clear 4 and 8 from inflights together.

Comment (xiang90) [inline raft/progress.go]: I am considering to just make this a direct bytes based control. But it requires some changes from the upper level rafthttp, which now has a message based buffer (N number of messages).

**Comment by ongardie** [comment]
@philips, I mentioned this one last night, I think you wanted to take a look

**Comment by gyuho** [comment]
@xiang90 Shall we merge this? Low risk. lgtm.

File: raft/progress.go [hunk 3]
```diff
@@ -183,14 +188,13 @@ type inflights struct {
 	// number of inflights in the buffer
 	count int
 
-	// the size of the buffer
-	size   int
+	// buffer contains the index of the last entry
```
Comment (gyuho) [inline raft/progress.go]: Would it be better to word

```
// buffer contains the last entry indexes of each message.
```

**Comment by gyuho** [comment]
@xiang90 I think this is conflict with https://github.com/coreos/etcd/pull/6353.I can cherry-pick comment part if you want.

**Pull request closed without merging**